# Fase 3 — Regras de Associação: Candidatos 2026 (TSE)

## Pergunta de negócio

> **Que combinações de características dos candidatos aparecem juntas com mais frequência do que se fossem independentes? Em particular: quais atributos (partido, gênero, escolaridade, cor/raça, UF de nascimento) estão associados a cada perfil identificado na Fase 2 — e, assim que a eleição ocorrer, a ser eleito?**

Até aqui, a clusterização (Fase 2) agrupou candidatos por **distância** em variáveis numéricas (idade, escolaridade, patrimônio). Regras de associação fazem uma pergunta diferente: em vez de "quem está perto de quem", **quais combinações de atributos categóricos aparecem juntas com mais frequência do que o acaso explicaria**? A técnica clássica pra isso é o **Apriori**, que originalmente resolve o problema da "cesta de compras" (que produtos são comprados juntos) — aqui, cada candidato é uma cesta, e cada valor de cada variável categórica (partido, gênero, escolaridade...) é um item dessa cesta.


## Configuração

Use os **mesmos valores** que você usou nas Fases 0/1/2, para carregar o arquivo certo.


In [1]:
CARGO = "DEPUTADO FEDERAL"   # mesmo valor usado nas Fases 0/1/2
UF = 'AL', 'BA', 'CE', 'MA', 'PB', 'PE', 'PI', 'RN', 'SE'               # mesmo valor usado nas Fases 0/1/2
ANO_ELEICAO = 2026


## Preparando a base — incorporando os 4 clusters da Fase 2 e discretização em quartis

O algoritmo Apriori trabalha exclusivamente com **itens categóricos e binários** (presente/ausente). Para integrar a inteligência descoberta na Fase 2 e testar formalmente se os nomes dos clusters fazem sentido:
1. Carregamos diretamente os **4 arquétipos eleitorais** da Fase 2 (`Estruturados`, `Financiados`, `Base Simbólica`, `Patrimônio Sem Campanha`).
2. Discretizamos as **4 variáveis contínuas** utilizadas na clusterização em **quartis**:
   - **Idade** → 4 faixas por quartis puros (`pd.qcut`).
   - **Anos de Estudo** → 4 faixas por quartis (com `duplicates='drop'` caso necessário).
   - **Total de Bens** → tratamento *zero-aware*: zeros separados como `Sem Bens (R$ 0)` + 3 quartis sobre os valores positivos.
   - **Total de Despesas Contratadas** → tratamento *zero-aware*: zeros separados como `Sem Despesa (R$ 0)` + 3 quartis sobre os valores positivos.

> **Por que quartis *zero-aware*?**  Cerca de 43% dos candidatos declararam bens = R$ 0 e 26% registraram despesas = R$ 0. Se aplicássemos `pd.qcut` puro, os dois primeiros quartis colapsariam em zero, perdendo poder discriminativo. A abordagem adotada preserva a categoria "zero" como informação analítica e calcula quartis apenas sobre quem de fato possui valores positivos.

In [2]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from mlxtend.frequent_patterns import apriori, association_rules
from pyvis.network import Network
from IPython.display import IFrame

sns.set_style("whitegrid")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)  # não truncar antecedents/consequents nas tabelas

SEMENTE = 42
np.random.seed(SEMENTE)  # mesmo racional de reprodutibilidade defensiva dos notebooks anteriores

nome_uf = UF if UF is not None else 'BRASIL'
nome_cargo = CARGO.replace(' ', '_')
arquivo = f"dados/candidatos_{nome_cargo}_nordeste_{ANO_ELEICAO}.csv"

df = pd.read_csv(arquivo)
print(f"Carregado: {arquivo}  ->  {df.shape[0]} candidatos, {df.shape[1]} colunas")
df.head()


Carregado: dados/candidatos_DEPUTADO_FEDERAL_nordeste_2026.csv  ->  2189 candidatos, 71 colunas


,DT_GERACAO,HH_GERACAO,ANO_ELEICAO,CD_TIPO_ELEICAO,NM_TIPO_ELEICAO,NR_TURNO,CD_ELEICAO,DS_ELEICAO,DT_ELEICAO,TP_ABRANGENCIA,SG_UF,SG_UE,NM_UE,CD_CARGO,DS_CARGO,SQ_CANDIDATO,NR_CANDIDATO,NM_CANDIDATO,NM_URNA_CANDIDATO,NM_SOCIAL_CANDIDATO,NR_CPF_CANDIDATO,DS_EMAIL,CD_SITUACAO_CANDIDATURA,DS_SITUACAO_CANDIDATURA,TP_AGREMIACAO,NR_PARTIDO,SG_PARTIDO,NM_PARTIDO,NR_FEDERACAO,NM_FEDERACAO,SG_FEDERACAO,DS_COMPOSICAO_FEDERACAO,SQ_COLIGACAO,NM_COLIGACAO,DS_COMPOSICAO_COLIGACAO,SG_UF_NASCIMENTO,DT_NASCIMENTO,NR_TITULO_ELEITORAL_CANDIDATO,CD_GENERO,DS_GENERO,CD_GRAU_INSTRUCAO,DS_GRAU_INSTRUCAO,CD_ESTADO_CIVIL,DS_ESTADO_CIVIL,CD_COR_RACA,DS_COR_RACA,CD_OCUPACAO,DS_OCUPACAO,CD_SIT_TOT_TURNO,DS_SIT_TOT_TURNO,BENS IMÓVEIS,BENS MÓVEIS,DINHEIRO,INVESTIMENTOS RENDA FIXA,INVESTIMENTOS RENDA VARIÁVEL,OUTROS BENS,Total_Bens,Total_Bens_Log,Total_Despesas_Contratadas,Total_Despesas_Pagas,Total_Despesas_Contratadas_Log,Total_Despesas_Pagas_Log,ANOS_ESTUDO,IDADE,cluster_v3_letra,cluster_provisorio,cluster_v2_letra,cluster_v2_nome,cluster_v3_nome,cluster_4v_nome,cluster
0,23/08/2026,19:30:42,2026,2,ELEIÇÃO ORDINÁRIA,1,6259,Eleições Gerais Estaduais 2026,2026-10-04,ESTADUAL,AL,AL,ALAGOAS,6,DEPUTADO FEDERAL,20002532430,3033,CAROLAYNNE PIRES DE OLIVEIRA SOUZA,CAROL SOUZA,#NULO,10279565488,NÃO DIVULGÁVEL,-3,#NE,PARTIDO ISOLADO,30,NOVO,PARTIDO NOVO,-1,#NULO,#NULO,#NULO,20001800111,PARTIDO ISOLADO,NOVO,AL,1991-11-10,40248081791,4,FEMININO,8,SUPERIOR COMPLETO,1,SOLTEIRO(A),2,PRETA,132,PSICÓLOGO,-1,#NULO,NaN,NaN,NaN,NaN,NaN,NaN,0.00,0.000000,15500.00,15500.00,9.648660,9.648660,16,34,C,Cluster C,L,NaN,Financiados,Financiados,Financiados
1,23/08/2026,19:30:42,2026,2,ELEIÇÃO ORDINÁRIA,1,6259,Eleições Gerais Estaduais 2026,2026-10-04,ESTADUAL,AL,AL,ALAGOAS,6,DEPUTADO FEDERAL,20002532431,3030,JOSÉ GUIDO DO RÊGO SANTOS NETO,GUIDO SANTOS,#NULO,5955901442,NÃO DIVULGÁVEL,-3,#NE,PARTIDO ISOLADO,30,NOVO,PARTIDO NOVO,-1,#NULO,#NULO,#NULO,20001800111,PARTIDO ISOLADO,NOVO,AL,1985-06-07,31724561767,2,MASCULINO,8,SUPERIOR COMPLETO,3,CASADO(A),1,BRANCA,296,SERVIDOR PÚBLICO FEDERAL,-1,#NULO,0.0,0.0,45000.00,0.0,0.0,70000.0,115000.00,11.652696,14099.27,14099.27,9.553949,9.553949,16,41,A,Cluster A,R,NaN,Estruturados,Estruturados,Estruturados
2,23/08/2026,19:30:42,2026,2,ELEIÇÃO ORDINÁRIA,1,6259,Eleições Gerais Estaduais 2026,2026-10-04,ESTADUAL,AL,AL,ALAGOAS,6,DEPUTADO FEDERAL,20002532432,3003,JOÃO CARLOS LIMA UCHÔA,JOÃO UCHÔA,#NULO,48288586449,NÃO DIVULGÁVEL,-3,#NE,PARTIDO ISOLADO,30,NOVO,PARTIDO NOVO,-1,#NULO,#NULO,#NULO,20001800111,PARTIDO ISOLADO,NOVO,AL,1965-02-12,5104321775,2,MASCULINO,8,SUPERIOR COMPLETO,3,CASADO(A),1,BRANCA,131,ADVOGADO,-1,#NULO,1850000.0,841018.0,77464.01,0.0,0.0,1000000.0,3768482.01,15.142183,19635.06,19635.06,9.885123,9.885123,16,61,A,Cluster A,K,NaN,Estruturados,Estruturados,Estruturados
3,23/08/2026,19:30:42,2026,2,ELEIÇÃO ORDINÁRIA,1,6259,Eleições Gerais Estaduais 2026,2026-10-04,ESTADUAL,AL,AL,ALAGOAS,6,DEPUTADO FEDERAL,20002532433,3000,ANA CAROLINA BELTRÃO PEIXOTO,PROFESSORA CAROL,#NULO,2734264439,NÃO DIVULGÁVEL,-3,#NE,PARTIDO ISOLADO,30,NOVO,PARTIDO NOVO,-1,#NULO,#NULO,#NULO,20001800111,PARTIDO ISOLADO,NOVO,AL,1978-09-21,24962281775,4,FEMININO,8,SUPERIOR COMPLETO,5,VIÚVO(A),1,BRANCA,297,SERVIDOR PÚBLICO ESTADUAL,-1,#NULO,570000.0,60000.0,0.00,0.0,0.0,0.0,630000.00,13.353477,51044.66,51043.66,10.840476,10.840456,16,48,A,Cluster A,I,NaN,Estruturados,Estruturados,Estruturados
4,23/08/2026,19:30:42,2026,2,ELEIÇÃO ORDINÁRIA,1,6259,Eleições Gerais Estaduais 2026,2026-10-04,ESTADUAL,AL,AL,ALAGOAS,6,DEPUTADO FEDERAL,20002532434,3001,JOSÉ ALEX BRANDÃO SILVEIRA,ALEX BRANDÃO,#NULO,3732710424,NÃO DIVULGÁVEL,-3,#NE,PARTIDO ISOLADO,30,NOVO,PARTIDO NOVO,-1,#NULO,#NULO,#NULO,20001800111,PARTIDO ISOLADO,NOVO,AL,1981-08-03,26986421767,2,MASCULINO,8,SUPERIOR COMPLETO,3,CASADO(A),1,BRANCA,403,"CORRETOR DE IMÓVEIS, SEGUROS, TÍTULOS E VALORES",-1,#NULO,950000.0,0.0,0.00,0.0,0.0,142000.0,1092000.00,13.903522,21932.67,21932.67,9.995778,9.995778,16,45,A,Cluster A,I,NaN,Est

In [3]:
# 1. Validação e atribuição do cluster dos candidatos
if 'cluster_4v_nome' in df.columns:
    df['cluster'] = df['cluster_4v_nome']
elif 'cluster_v3_nome' in df.columns:
    df['cluster'] = df['cluster_v3_nome']
else:
    # Fallback defensivo de reconstituição do modelo de 4 drivers da Fase 2
    colunas_driver = ['IDADE', 'ANOS_ESTUDO', 'Total_Bens_Log', 'Total_Despesas_Contratadas_Log']
    scaler = MinMaxScaler()
    X = scaler.fit_transform(df[colunas_driver])
    km = KMeans(n_clusters=4, random_state=SEMENTE, n_init=10)
    map_v3 = {'A': 'Estruturados', 'B': 'Base Simbólica', 'C': 'Financiados', 'D': 'Patrimônio Sem Campanha'}
    df['cluster'] = [map_v3[chr(65 + x)] for x in km.fit_predict(X)]

# ── Discretização das 4 variáveis contínuas em QUARTIS ──────────────────────
# Conforme orientação do professor: variáveis contínuas devem ser transformadas
# em categorias por meio de quartis para uso no Apriori.

# 2a. IDADE — quartis puros (variável sem zeros)
df['Faixa_Idade'] = pd.qcut(
    df['IDADE'], q=4, duplicates='drop',
    labels=['Idade Q1 (mais jovens)', 'Idade Q2', 'Idade Q3', 'Idade Q4 (mais velhos)']
)
print('Quartis de IDADE:')
display(df['Faixa_Idade'].value_counts().sort_index().to_frame('Qtd'))

# 2b. ANOS_ESTUDO — discretização baseada nos graus de instrução
# A variável tem concentração extrema: ~58% dos candidatos possuem Superior Completo (16 anos).
# Q2 = Q3 = 16, o que colapsa pd.qcut em apenas 2 faixas. Usamos cortes interpretativos
# alinhados aos níveis reais de escolaridade, mantendo o espírito de quartis:
#   Até Fundamental (≤ 8 anos) | Médio Incompleto/Completo (9-11) | Superior Incompleto (12-13) | Superior Completo (≥ 14)
bins_estudo = [-np.inf, 8, 11, 13, np.inf]
labels_estudo = [
    'Estudo: Até Fundamental',
    'Estudo: Ensino Médio',
    'Estudo: Superior Incompleto',
    'Estudo: Superior Completo'
]
df['Faixa_Estudo'] = pd.cut(df['ANOS_ESTUDO'], bins=bins_estudo, labels=labels_estudo, include_lowest=True)
print('\nQuartis de ANOS_ESTUDO:')
display(df['Faixa_Estudo'].value_counts().sort_index().to_frame('Qtd'))

# 2c. TOTAL DE BENS — zero-aware quartiles
# ~43% dos candidatos declararam bens = R$ 0; separamos os zeros como categoria própria
mask_bens = df['Total_Bens'] > 0
df['Faixa_Bens'] = 'Sem Bens (R$ 0)'
quartis_bens = pd.qcut(
    df.loc[mask_bens, 'Total_Bens'], q=3, duplicates='drop',
    labels=['Bens Q1 (baixo)', 'Bens Q2 (médio)', 'Bens Q3 (alto)']
)
df.loc[mask_bens, 'Faixa_Bens'] = quartis_bens.astype(str)
print('\nQuartis de Total_Bens (zero-aware):')
display(df['Faixa_Bens'].value_counts().to_frame('Qtd'))

# 2d. TOTAL DE DESPESAS CONTRATADAS — zero-aware quartiles
# ~26% registraram despesas = R$ 0; mesma estratégia
mask_desp = df['Total_Despesas_Contratadas'] > 0
df['Faixa_Despesa_Contratada'] = 'Sem Despesa (R$ 0)'
quartis_desp = pd.qcut(
    df.loc[mask_desp, 'Total_Despesas_Contratadas'], q=3, duplicates='drop',
    labels=['Despesa Q1 (baixa)', 'Despesa Q2 (média)', 'Despesa Q3 (alta)']
)
df.loc[mask_desp, 'Faixa_Despesa_Contratada'] = quartis_desp.astype(str)
print('\nQuartis de Total_Despesas_Contratadas (zero-aware):')
display(df['Faixa_Despesa_Contratada'].value_counts().to_frame('Qtd'))

# ── Resumo final ────────────────────────────────────────────────────────────
df_cluster = df.copy()
print('\n' + '='*60)
print('Distribuição dos 4 Clusters da Fase 2:')
display(df_cluster['cluster'].value_counts().to_frame('Qtd Candidatos'))

Quartis de IDADE:


,Qtd
Faixa_Idade,
Idade Q1 (mais jovens),567
Idade Q2,595
Idade Q3,485
Idade Q4 (mais velhos),542



Quartis de ANOS_ESTUDO:


,Qtd
Faixa_Estudo,
Estudo: Até Fundamental,121
Estudo: Ensino Médio,538
Estudo: Superior Incompleto,253
Estudo: Superior Completo,1277



Quartis de Total_Bens (zero-aware):


,Qtd
Faixa_Bens,
Sem Bens (R$ 0),771
Bens Q1 (baixo),474
Bens Q3 (alto),473
Bens Q2 (médio),471



Quartis de Total_Despesas_Contratadas (zero-aware):


,Qtd
Faixa_Despesa_Contratada,
Sem Despesa (R$ 0),557
Despesa Q1 (baixa),544
Despesa Q2 (média),544
Despesa Q3 (alta),544



Distribuição dos 4 Clusters da Fase 2:


,Qtd Candidatos
cluster,
Estruturados,1153
Financiados,460
Base Simbólica,328
Patrimônio Sem Campanha,248


### Base consolidada para mineração

Todos os 2.189 candidatos do Nordeste são mantidos na análise, com seus respectivos clusters multidimensionais e faixas de gastos contratados devidamente tipados.

In [4]:
display(df_cluster[['NM_CANDIDATO', 'SG_UF', 'cluster', 'Faixa_Bens', 'Faixa_Despesa_Contratada', 'DS_OCUPACAO']].head(8))

,NM_CANDIDATO,SG_UF,cluster,Faixa_Bens,Faixa_Despesa_Contratada,DS_OCUPACAO
0,CAROLAYNNE PIRES DE OLIVEIRA SOUZA,AL,Financiados,Sem Bens (R$ 0),Despesa Q1 (baixa),PSICÓLOGO
1,JOSÉ GUIDO DO RÊGO SANTOS NETO,AL,Estruturados,Bens Q1 (baixo),Despesa Q1 (baixa),SERVIDOR PÚBLICO FEDERAL
2,JOÃO CARLOS LIMA UCHÔA,AL,Estruturados,Bens Q3 (alto),Despesa Q1 (baixa),ADVOGADO
3,ANA CAROLINA BELTRÃO PEIXOTO,AL,Estruturados,Bens Q2 (médio),Despesa Q2 (média),SERVIDOR PÚBLICO ESTADUAL
4,JOSÉ ALEX BRANDÃO SILVEIRA,AL,Estruturados,Bens Q3 (alto),Despesa Q1 (baixa),"CORRETOR DE IMÓVEIS, SEGUROS, TÍTULOS E VALORES"
5,SUZICLEI PEREIRA DOS SANTOS,AL,Financiados,Sem Bens (R$ 0),Despesa Q1 (baixa),ASSISTENTE SOCIAL
6,GERVASIO RAIMUNDO DOS SANTOS NETO,AL,Patrimônio Sem Campanha,Bens Q1 (baixo),Sem Despesa (R$ 0),ADVOGADO
7,JOAO FRANCISCO DE ASSIS NETO,AL,Base Simbólica,Sem Bens (R$ 0),Sem Despesa (R$ 0),ADVOGADO


## Montando a "cesta" de itens

O algoritmo Apriori requer **variáveis booleanas** (one-hot encoding). Cada candidato corresponde a uma "transação" e cada atributo categórico vira um item (`SG_PARTIDO=PT`, `cluster=Estruturados`, `Faixa_Despesa_Contratada=Alta (> R$ 150k)`, `DS_OCUPACAO=DEPUTADO`, etc.).

Adicionamos a **ocupação declarada (`DS_OCUPACAO`)** e a **faixa de despesa contratada**, ampliando a profundidade socioprofissional e financeira das regras.

In [5]:
colunas_apriori = [
    'SG_PARTIDO',
    'DS_GENERO',
    'DS_GRAU_INSTRUCAO',
    'DS_ESTADO_CIVIL',
    'DS_COR_RACA',
    'cluster',
    'Faixa_Bens',
    'Faixa_Despesa_Contratada',
    'Faixa_Idade',       # quartis de idade
    'Faixa_Estudo',      # quartis de anos de estudo
    'DS_OCUPACAO'
]

if UF is None:
    colunas_apriori.append('SG_UF')

# Se o resultado da eleição estiver disponível
usar_resultado_eleicao = df_cluster['DS_SIT_TOT_TURNO'].nunique() > 1
if usar_resultado_eleicao:
    colunas_apriori.append('DS_SIT_TOT_TURNO')

cesta_base = df_cluster[['SQ_CANDIDATO'] + colunas_apriori].copy()
cesta = pd.get_dummies(cesta_base, columns=colunas_apriori, prefix_sep='=')
cesta = cesta.set_index('SQ_CANDIDATO')

print(f'{cesta.shape[0]} candidatos (cestas) x {cesta.shape[1]} itens possíveis (um por valor de cada variável)')
cesta_base.head()

2189 candidatos (cestas) x 206 itens possíveis (um por valor de cada variável)


,SQ_CANDIDATO,SG_PARTIDO,DS_GENERO,DS_GRAU_INSTRUCAO,DS_ESTADO_CIVIL,DS_COR_RACA,cluster,Faixa_Bens,Faixa_Despesa_Contratada,Faixa_Idade,Faixa_Estudo,DS_OCUPACAO
0,20002532430,NOVO,FEMININO,SUPERIOR COMPLETO,SOLTEIRO(A),PRETA,Financiados,Sem Bens (R$ 0),Despesa Q1 (baixa),Idade Q1 (mais jovens),Estudo: Superior Completo,PSICÓLOGO
1,20002532431,NOVO,MASCULINO,SUPERIOR COMPLETO,CASADO(A),BRANCA,Estruturados,Bens Q1 (baixo),Despesa Q1 (baixa),Idade Q2,Estudo: Superior Completo,SERVIDOR PÚBLICO FEDERAL
2,20002532432,NOVO,MASCULINO,SUPERIOR COMPLETO,CASADO(A),BRANCA,Estruturados,Bens Q3 (alto),Despesa Q1 (baixa),Idade Q4 (mais velhos),Estudo: Superior Completo,ADVOGADO
3,20002532433,NOVO,FEMININO,SUPERIOR COMPLETO,VIÚVO(A),BRANCA,Estruturados,Bens Q2 (médio),Despesa Q2 (média),Idade Q2,Estudo: Superior Completo,SERVIDOR PÚBLICO ESTADUAL
4,20002532434,NOVO,MASCULINO,SUPERIOR COMPLETO,CASADO(A),BRANCA,Estruturados,Bens Q3 (alto),Despesa Q1 (baixa),Idade Q2,Estudo: Superior Completo,"CORRETOR DE IMÓVEIS, SEGUROS, TÍTULOS E VALORES"


## Itens frequentes: o que aparece sozinho, com que frequência

O Apriori funciona em duas etapas. Primeiro, encontra **itemsets frequentes** — combinações de 1, 2, 3... itens que aparecem juntas em pelo menos `min_suporte` das cestas. Só depois, na segunda etapa, ele vira **regras** (`A → B`). Começamos pelos itemsets de um item só — a frequência "crua" de cada valor, antes de cruzar qualquer coisa.


In [6]:
# Suporte mínimo calibrado para 1.5% (~33 candidatos), permitindo capturar ocupações relevantes
min_suporte = 0.015

itens_frequentes = apriori(cesta, min_support=min_suporte, use_colnames=True)
print(f'{len(itens_frequentes)} itemsets frequentes (min_suporte={min_suporte})')

individuais = itens_frequentes[itens_frequentes['itemsets'].apply(lambda x: len(x) == 1)].copy()
individuais['item'] = individuais['itemsets'].apply(lambda x: next(iter(x)))
individuais['qtd_candidatos'] = (individuais['support'] * len(cesta)).round().astype(int)
display(individuais.sort_values('support', ascending=False)[['item', 'support', 'qtd_candidatos']].head(20))


8231 itemsets frequentes (min_suporte=0.015)


,item,support,qtd_candidatos
21,DS_GENERO=MASCULINO,0.623572,1365
26,DS_GRAU_INSTRUCAO=SUPERIOR COMPLETO,0.583371,1277
53,Faixa_Estudo=Estudo: Superior Completo,0.583371,1277
35,cluster=Estruturados,0.526725,1153
28,DS_ESTADO_CIVIL=CASADO(A),0.490635,1074
32,DS_COR_RACA=PARDA,0.472362,1034
20,DS_GENERO=FEMININO,0.376428,824
30,DS_ESTADO_CIVIL=SOLTEIRO(A),0.356784,781
41,Faixa_Bens=Sem Bens (R$ 0),0.352216,771
31,DS_COR_RACA=BRANCA,0.350845,768


### Suporte, confiança e lift — o que cada métrica mede

Regras de associação têm o formato **"se A, então B"** (A = antecedente, B = consequente), e três números resumem o quão forte é essa relação:

- **Suporte** — fração de candidatos em que A e B aparecem juntos: `suporte(A→B) = P(A e B)`. Mede o quão comum é a combinação; suporte baixo (poucos candidatos) é terreno fértil pra coincidência.
- **Confiança** — entre os candidatos que têm A, qual fração também tem B: `confiança(A→B) = P(B | A) = suporte(A e B) / suporte(A)`. É a "taxa de acerto" da regra.
- **Lift** — o quanto a confiança da regra é maior (ou menor) do que se A e B fossem independentes: `lift(A→B) = confiança(A→B) / suporte(B)`. `lift = 1` → A e B são independentes (a regra não diz nada); `lift > 1` → aparecem juntos mais do que o esperado ao acaso; `lift < 1` → aparecem juntos menos do que o esperado.

**Um cuidado importante nesta base:** ela tem só ~200 candidatos espalhados por 27 partidos e 27 estados — em média, uns 7 candidatos por partido ou por UF. Com grupos tão pequenos, bastam 3 ou 4 coincidências pra um lift parecer altíssimo, sem que isso signifique relação real nenhuma. Por isso toda tabela de regras abaixo vem com uma coluna `qtd_candidatos` — sempre olhem ela antes de confiar num lift alto.


In [7]:
regras = association_rules(itens_frequentes, metric='lift', min_threshold=1.1)
regras['qtd_candidatos'] = (regras['support'] * len(cesta)).round().astype(int)
regras = regras.sort_values('confidence', ascending=False)

def formatar_regras(df_regras):
    df_fmt = df_regras.copy()
    df_fmt['antecedents'] = df_fmt['antecedents'].apply(lambda x: ', '.join(sorted(str(i) for i in x)))
    df_fmt['consequents'] = df_fmt['consequents'].apply(lambda x: ', '.join(sorted(str(i) for i in x)))
    return df_fmt

colunas_exibir = ['antecedents', 'consequents', 'qtd_candidatos', 'support', 'confidence', 'lift']
print(f'{len(regras)} regras encontradas com lift >= 1.1')
display(formatar_regras(regras[colunas_exibir].head(10)))


126232 regras encontradas com lift >= 1.1


,antecedents,consequents,qtd_candidatos,support,confidence,lift
59516,"DS_ESTADO_CIVIL=SOLTEIRO(A), DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Despesa_Contratada=Despesa Q1 (baixa), cluster=Financiados",Faixa_Estudo=Estudo: Ensino Médio,33,0.015075,1.0,4.068773
59486,"DS_ESTADO_CIVIL=SOLTEIRO(A), DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Bens=Sem Bens (R$ 0), cluster=Financiados",Faixa_Estudo=Estudo: Ensino Médio,58,0.026496,1.0,4.068773
15485,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Despesa_Contratada=Despesa Q2 (média)",Faixa_Estudo=Estudo: Ensino Médio,67,0.030608,1.0,4.068773
41934,"DS_ESTADO_CIVIL=CASADO(A), DS_GENERO=FEMININO, Faixa_Bens=Bens Q3 (alto), Faixa_Despesa_Contratada=Despesa Q3 (alta)",cluster=Estruturados,33,0.015075,1.0,1.898526
79909,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Bens=Sem Bens (R$ 0), Faixa_Despesa_Contratada=Sem Despesa (R$ 0), Faixa_Estudo=Estudo: Ensino Médio",cluster=Base Simbólica,71,0.032435,1.0,6.673780
79912,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Bens=Sem Bens (R$ 0), Faixa_Despesa_Contratada=Sem Despesa (R$ 0), cluster=Base Simbólica",Faixa_Estudo=Estudo: Ensino Médio,71,0.032435,1.0,4.068773
79924,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO, Faixa_Bens=Sem Bens (R$ 0), Faixa_Despesa_Contratada=Sem Despesa (R$ 0)","Faixa_Estudo=Estudo: Ensino Médio, cluster=Base Simbólica",71,0.032435,1.0,15.524823
27802,"Faixa_Despesa_Contratada=Despesa Q1 (baixa), Faixa_Estudo=Estudo: Superior Completo, Faixa_Idade=Idade Q1 (mais jovens)",DS_GRAU_INSTRUCAO=SUPERIOR COMPLETO,77,0.035176,1.0,1.714174
27803,"DS_GRAU_INSTRUCAO=SUPERIOR COMPLETO, Faixa_Despesa_Contratada=Despesa Q1 (baixa), Faixa_Idade=Idade Q1 (mais jovens)",Faixa_Estudo=Estudo: Superior Completo,77,0.035176,1.0,1.714174
96309,"DS_COR_RACA=BRANCA, DS_ESTADO_CIVIL=CASADO(A), Faixa_Estudo=Estudo: Superior Completo, Faixa_Idade=Idade Q3, cluster=Estruturados",DS_GRAU_INSTRUCAO=SUPERIOR COMPLETO,45,0.020557,1.0,1.714174


### Cuidado: suporte pequeno infla o lift

Vamos ver isso na prática: as regras de maior lift da base inteira, sem nenhum filtro de suporte, comparadas às regras de maior lift **exigindo pelo menos 10 candidatos** por trás delas.


In [8]:
regras_1_consequente = regras[regras['consequents'].apply(lambda x: len(x) == 1)].copy()
colunas_exibir = ['antecedents', 'consequents', 'qtd_candidatos', 'confidence', 'lift']

print("--- Maior lift, sem filtro de suporte ---")
display(formatar_regras(regras_1_consequente.sort_values('lift', ascending=False)[colunas_exibir].head(5)))

print("--- Maior lift, exigindo qtd_candidatos >= 10 ---")
display(formatar_regras(regras_1_consequente[regras_1_consequente['qtd_candidatos'] >= 10]
        .sort_values('lift', ascending=False)[colunas_exibir].head(5)))


--- Maior lift, sem filtro de suporte ---


,antecedents,consequents,qtd_candidatos,confidence,lift
3599,"Faixa_Bens=Sem Bens (R$ 0), Faixa_Estudo=Estudo: Até Fundamental",DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO,39,0.629032,21.856375
2724,"DS_GENERO=MASCULINO, Faixa_Estudo=Estudo: Até Fundamental",DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL COMPLETO,34,0.395349,18.413162
2728,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO",Faixa_Estudo=Estudo: Até Fundamental,43,1.000000,18.090909
258,DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL COMPLETO,Faixa_Estudo=Estudo: Até Fundamental,47,1.000000,18.090909
3592,"DS_COR_RACA=PARDA, DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO",Faixa_Estudo=Estudo: Até Fundamental,35,1.000000,18.090909


--- Maior lift, exigindo qtd_candidatos >= 10 ---


,antecedents,consequents,qtd_candidatos,confidence,lift
3599,"Faixa_Bens=Sem Bens (R$ 0), Faixa_Estudo=Estudo: Até Fundamental",DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO,39,0.629032,21.856375
2724,"DS_GENERO=MASCULINO, Faixa_Estudo=Estudo: Até Fundamental",DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL COMPLETO,34,0.395349,18.413162
2728,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO",Faixa_Estudo=Estudo: Até Fundamental,43,1.000000,18.090909
258,DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL COMPLETO,Faixa_Estudo=Estudo: Até Fundamental,47,1.000000,18.090909
3592,"DS_COR_RACA=PARDA, DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO",Faixa_Estudo=Estudo: Até Fundamental,35,1.000000,18.090909


No primeiro grupo, é comum ver pares como `SG_UF=ES → SG_UF_NASCIMENTO=ES` com lift acima de 30: candidatos a governador costumam concorrer no estado onde nasceram, então UF e UF de nascimento são quase a mesma informação — com só 4 ou 5 candidatos capixabas na base, a coincidência vira um lift enorme sem dizer nada de novo. No segundo grupo, com um piso de candidatos por trás da regra, sobram combinações genuinamente mais interessantes (ex.: escolaridade + gênero prevendo cluster). Essa é a leitura que vamos seguir usando daqui pra frente.

### Filtrando pra regras que valem a pena olhar


In [9]:
suporte_pratico = 0.02    # ~44 candidatos
confianca_pratica = 0.4

regras_simples = regras[
    regras['antecedents'].apply(lambda x: len(x) <= 2) &
    regras['consequents'].apply(lambda x: len(x) == 1) &
    (regras['support'] >= suporte_pratico) &
    (regras['confidence'] >= confianca_pratica)
].sort_values('lift', ascending=False)

print(f'{len(regras_simples)} regras com suporte >= {suporte_pratico} e confiança >= {confianca_pratica}')
display(formatar_regras(regras_simples[colunas_exibir].head(15)))


1857 regras com suporte >= 0.02 e confiança >= 0.4


,antecedents,consequents,qtd_candidatos,confidence,lift
258,DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL COMPLETO,Faixa_Estudo=Estudo: Até Fundamental,47,1.000000,18.090909
264,DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO,Faixa_Estudo=Estudo: Até Fundamental,63,1.000000,18.090909
265,Faixa_Estudo=Estudo: Até Fundamental,DS_GRAU_INSTRUCAO=ENSINO FUNDAMENTAL INCOMPLETO,63,0.520661,18.090909
899,"Faixa_Despesa_Contratada=Despesa Q1 (baixa), Faixa_Idade=Idade Q1 (mais jovens)",SG_PARTIDO=MISSÃO,74,0.422857,9.542622
7698,"Faixa_Bens=Bens Q3 (alto), Faixa_Despesa_Contratada=Sem Despesa (R$ 0)",cluster=Patrimônio Sem Campanha,48,1.000000,8.826613
7691,"Faixa_Bens=Bens Q2 (médio), Faixa_Despesa_Contratada=Sem Despesa (R$ 0)",cluster=Patrimônio Sem Campanha,63,1.000000,8.826613
2890,"DS_GENERO=MASCULINO, DS_GRAU_INSTRUCAO=SUPERIOR INCOMPLETO",Faixa_Estudo=Estudo: Superior Incompleto,146,1.000000,8.652174
347,Faixa_Estudo=Estudo: Superior Incompleto,DS_GRAU_INSTRUCAO=SUPERIOR INCOMPLETO,253,1.000000,8.652174
5049,"Faixa_Estudo=Estudo: Superior Incompleto, Faixa_Idade=Idade Q3",DS_GRAU_INSTRUCAO=SUPERIOR INCOMPLETO,49,1.000000,8.652174
5053,"DS_GRAU_INSTRUCAO=SUPERIOR INCOMPLETO, DS_OCUPACAO=EMPRESÁRIO",Faixa_Estudo=Estudo: Superior Incompleto,55,1.000000,8.652174


## O que prediz o cluster de um candidato?

Filtramos as regras cujo **consequente é um único item de cluster** — ou seja, *"se o candidato tem tais características, com qual probabilidade e força de associação ele cai no cluster X?"*.

Esta é a pergunta-chave para testar: **a nossa nomeação dos clusters da Fase 2 faz sentido no mundo real?**
Se o Apriori encontrar regras com 100% de confiança e lift elevado associando cada cluster exatamente aos atributos conceituais que deram origem ao seu nome, a nomenclatura está plenamente validada.

In [10]:
def regras_para_consequente(regras_completas, prefixo, n_por_valor=6, min_suporte=min_suporte, max_antecedentes=2):
    candidatas = regras_completas[
        regras_completas['consequents'].apply(lambda x: len(x) == 1 and next(iter(x)).startswith(prefixo)) &
        regras_completas['antecedents'].apply(lambda x: len(x) <= max_antecedentes) &
        (regras_completas['support'] >= min_suporte)
    ].copy()
    candidatas['alvo'] = candidatas['consequents'].apply(lambda x: next(iter(x)))
    resultado = (candidatas.sort_values('lift', ascending=False)
                            .groupby('alvo', group_keys=False)
                            .head(n_por_valor)
                            .sort_values(['alvo', 'lift'], ascending=[True, False]))
    return resultado.drop(columns='alvo')

regras_cluster = regras_para_consequente(regras, 'cluster=')
print(f'{len(regras_cluster)} regras (até 6 por cluster, ordenadas por lift)')
display(formatar_regras(regras_cluster[colunas_exibir]))


24 regras (até 6 por cluster, ordenadas por lift)


,antecedents,consequents,qtd_candidatos,confidence,lift
7038,"Faixa_Bens=Sem Bens (R$ 0), Faixa_Despesa_Contratada=Sem Despesa (R$ 0)",cluster=Base Simbólica,308,1.000000,6.673780
688,"Faixa_Bens=Sem Bens (R$ 0), SG_PARTIDO=DC",cluster=Base Simbólica,46,0.867925,5.792338
694,"Faixa_Despesa_Contratada=Sem Despesa (R$ 0), SG_PARTIDO=DC",cluster=Base Simbólica,46,0.766667,5.116565
2410,"DS_GENERO=FEMININO, Faixa_Despesa_Contratada=Sem Despesa (R$ 0)",cluster=Base Simbólica,130,0.702703,4.689684
7112,"Faixa_Despesa_Contratada=Sem Despesa (R$ 0), Faixa_Estudo=Estudo: Até Fundamental",cluster=Base Simbólica,34,0.693878,4.630786
7138,"DS_OCUPACAO=OUTROS, Faixa_Despesa_Contratada=Sem Despesa (R$ 0)",cluster=Base Simbólica,59,0.678161,4.525897
7262,"Faixa_Bens=Bens Q3 (alto), Faixa_Despesa_Contratada=Despesa Q2 (média)",cluster=Estruturados,90,1.000000,1.898526
7267,"Faixa_Bens=Bens Q3 (alto), Faixa_Despesa_Contratada=Despesa Q3 (alta)",cluster=Estruturados,282,1.000000,1.898526
7172,"Faixa_Bens=Bens Q1 (baixo), Faixa_Despesa_Contratada=Despesa Q3 (alta)",cluster=Estruturados,71,1.000000,1.898526
7219,"Faixa_Bens=Bens Q2 (médio), Faixa_Despesa_Contratada=Despesa Q3 (alta)",cluster=Estruturados,142,1.000000,1.898526


### Visualizando como grafo

Uma tabela com dezenas de regras é difícil de escanear — um grafo interativo ajuda a ver de relance quais itens aparecem mais como "explicação" (antecedentes) e quais aparecem mais como "resultado" (consequentes). Cada item vira um nó (`SG_PARTIDO=PCO`, `cluster=D`, ...); cada regra vira uma seta do antecedente pro consequente, com espessura proporcional à confiança. Os nós que começam com `cluster=` ficam destacados em vermelho e maiores, pra serem fáceis de achar no meio dos outros itens. Passe o mouse sobre uma seta pra ver confiança, lift e quantos candidatos sustentam aquela regra — e arraste os nós se a rede ficar embaralhada.


In [11]:
def grafo_regras_html(regras_plot, arquivo_html, titulo, prefixo_destaque='cluster='):
    print(titulo)
    net = Network(height='650px', width='100%', bgcolor='#222222', font_color='white',
                  notebook=True, directed=True, cdn_resources='in_line')

    nos_adicionados = set()
    for _, row in regras_plot.iterrows():
        antecedente = ' + '.join(sorted(str(item) for item in row['antecedents']))
        consequente = ' + '.join(sorted(str(item) for item in row['consequents']))

        for label, itemset in [(antecedente, row['antecedents']), (consequente, row['consequents'])]:
            if label in nos_adicionados:
                continue
            destaque = any(str(item).startswith(prefixo_destaque) for item in itemset)
            net.add_node(
                label, label, title=label,
                color='#e41a1c' if destaque else '#1f78b4',
                size=30 if destaque else 14,
                font={'size': 20 if destaque else 13},
            )
            nos_adicionados.add(label)

        titulo_aresta = (f"Confiança: {row['confidence']:.2f} | Lift: {row['lift']:.2f} | "
                          f"{row['qtd_candidatos']} candidatos")
        net.add_edge(antecedente, consequente, value=row['confidence'], title=titulo_aresta)

    net.show_buttons(filter_=['physics'])
    net.save_graph(arquivo_html)
    print(f'Grafo salvo com sucesso em: {arquivo_html}')
    return IFrame(arquivo_html, width='100%', height=750)

grafo_regras_html(regras_cluster.head(25), 'grafo_regras_cluster.html', 'O que prediz o cluster de um candidato (Despesas Contratadas)?')


O que prediz o cluster de um candidato (Despesas Contratadas)?
Grafo salvo com sucesso em: grafo_regras_cluster.html


## O que caracteriza quem já é de um cluster?

Agora invertemos a perspectiva analítica: fixamos o **antecedente** (`cluster=X`) e perguntamos: *"dado que um candidato pertence ao Cluster X, quais outros atributos (partidos, ocupações, gênero, escolaridade) aparecem com lift elevado?"*.

Isso nos revela o perfil sociopolítico intrínseco de cada arquétipo eleitoral.

In [12]:
def regras_para_antecedente(regras_completas, item_fixo, n=8, min_suporte=min_suporte, max_consequentes=1):
    candidatas = regras_completas[
        regras_completas['antecedents'].apply(lambda x: len(x) == 1 and next(iter(x)) == item_fixo) &
        regras_completas['consequents'].apply(lambda x: len(x) <= max_consequentes) &
        (regras_completas['support'] >= min_suporte)
    ]
    return candidatas.sort_values('lift', ascending=False).head(n)

os_4_clusters = ['cluster=Estruturados', 'cluster=Financiados', 'cluster=Base Simbólica', 'cluster=Patrimônio Sem Campanha']
regras_por_cluster_fixo = {
    nome: regras_para_antecedente(regras, nome, n=8)
    for nome in os_4_clusters
}

for item_fixo, subset in regras_por_cluster_fixo.items():
    print(f'=== {item_fixo} ({len(subset)} regras) ===')
    display(formatar_regras(subset[colunas_exibir]))


=== cluster=Estruturados (8 regras) ===


,antecedents,consequents,qtd_candidatos,confidence,lift
489,cluster=Estruturados,DS_OCUPACAO=DEPUTADO,83,0.071986,1.832298
480,cluster=Estruturados,Faixa_Despesa_Contratada=Despesa Q3 (alta),495,0.429315,1.727519
477,cluster=Estruturados,Faixa_Bens=Bens Q3 (alto),424,0.367736,1.701850
475,cluster=Estruturados,Faixa_Bens=Bens Q2 (médio),405,0.351258,1.632490
107,cluster=Estruturados,SG_PARTIDO=PP,58,0.050304,1.619331
493,cluster=Estruturados,DS_OCUPACAO=MÉDICO,57,0.049436,1.591411
494,cluster=Estruturados,DS_OCUPACAO=VEREADOR,59,0.051171,1.555736
161,cluster=Estruturados,SG_PARTIDO=PT,75,0.065048,1.531069


=== cluster=Financiados (8 regras) ===


,antecedents,consequents,qtd_candidatos,confidence,lift
497,cluster=Financiados,Faixa_Bens=Sem Bens (R$ 0),454,0.986957,2.802137
499,cluster=Financiados,Faixa_Despesa_Contratada=Despesa Q1 (baixa),241,0.523913,2.108172
41,cluster=Financiados,SG_PARTIDO=MISSÃO,42,0.091304,2.060466
508,cluster=Financiados,DS_OCUPACAO=OUTROS,99,0.215217,1.670606
51,cluster=Financiados,SG_PARTIDO=NOVO,46,0.100000,1.552482
501,cluster=Financiados,Faixa_Despesa_Contratada=Despesa Q2 (média),170,0.369565,1.487092
335,cluster=Financiados,DS_GRAU_INSTRUCAO=SUPERIOR INCOMPLETO,74,0.160870,1.391871
507,cluster=Financiados,Faixa_Estudo=Estudo: Superior Incompleto,74,0.160870,1.391871


=== cluster=Base Simbólica (8 regras) ===


,antecedents,consequents,qtd_candidatos,confidence,lift
17,cluster=Base Simbólica,SG_PARTIDO=DC,46,0.140244,4.514616
461,cluster=Base Simbólica,Faixa_Despesa_Contratada=Sem Despesa (R$ 0),319,0.972561,3.822147
459,cluster=Base Simbólica,Faixa_Bens=Sem Bens (R$ 0),317,0.966463,2.743954
465,cluster=Base Simbólica,Faixa_Estudo=Estudo: Até Fundamental,35,0.106707,1.930432
467,cluster=Base Simbólica,Faixa_Estudo=Estudo: Ensino Médio,141,0.429878,1.749076
273,cluster=Base Simbólica,DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO,119,0.362805,1.657996
471,cluster=Base Simbólica,DS_OCUPACAO=OUTROS,59,0.179878,1.396287
391,cluster=Base Simbólica,DS_ESTADO_CIVIL=SOLTEIRO(A),157,0.478659,1.341592


=== cluster=Patrimônio Sem Campanha (8 regras) ===


,antecedents,consequents,qtd_candidatos,confidence,lift
515,cluster=Patrimônio Sem Campanha,Faixa_Despesa_Contratada=Sem Despesa (R$ 0),238,0.959677,3.771515
510,cluster=Patrimônio Sem Campanha,Faixa_Bens=Bens Q1 (baixo),133,0.536290,2.476666
276,cluster=Patrimônio Sem Campanha,DS_GRAU_INSTRUCAO=ENSINO MÉDIO COMPLETO,72,0.290323,1.326756
519,cluster=Patrimônio Sem Campanha,Faixa_Estudo=Estudo: Ensino Médio,78,0.314516,1.279695
512,cluster=Patrimônio Sem Campanha,Faixa_Bens=Bens Q2 (médio),66,0.266129,1.236850
238,cluster=Patrimônio Sem Campanha,DS_GENERO=MASCULINO,188,0.758065,1.215680
430,cluster=Patrimônio Sem Campanha,DS_COR_RACA=PARDA,134,0.540323,1.143874
516,cluster=Patrimônio Sem Campanha,Faixa_Idade=Idade Q4 (mais velhos),70,0.282258,1.139968


### Visualizando como grafo (antecedente fixo)

Mesmo grafo interativo de antes, só que agora as três "causas" (os clusters) são o ponto de partida fixo, e os itens que aparecem como consequência é que variam.


In [13]:
regras_antecedente_fixo = pd.concat(regras_por_cluster_fixo.values())
grafo_regras_html(regras_antecedente_fixo, 'grafo_regras_antecedente_fixo.html', 'O que caracteriza quem pertence a cada um dos 4 clusters?')


O que caracteriza quem pertence a cada um dos 4 clusters?
Grafo salvo com sucesso em: grafo_regras_antecedente_fixo.html


## E quando a eleição já tiver ocorrido?

Esta seção só roda depois que a Fase 0 for rodada de novo com o resultado oficial (`DS_SIT_TOT_TURNO` deixa de vir toda com o mesmo valor `#NULO`) — por enquanto, ela se anuncia e não faz nada. Repare que não precisamos escrever nenhum código novo: como `DS_SIT_TOT_TURNO` já entrou na cesta lá na preparação (quando aplicável), a mesma `regras_para_consequente` e o mesmo `grafo_regras_html` servem pra essa pergunta também.


In [14]:
if usar_resultado_eleicao:
    regras_eleicao = regras_para_consequente(regras, 'DS_SIT_TOT_TURNO=')
    print(f"{len(regras_eleicao)} regras apontando para o resultado da eleição")
    display(formatar_regras(regras_eleicao[colunas_exibir].head(15)))
    grafo_regras_html(regras_eleicao.head(20), 'grafo_regras_eleicao.html', 'O que prediz ser eleito?', prefixo_destaque='DS_SIT_TOT_TURNO=')
else:
    print("DS_SIT_TOT_TURNO ainda não tem informação real (a eleição de 2026 não ocorreu) — "
          "essa seção fica pronta pra usar assim que a Fase 0 for rodada de novo com o resultado oficial. "
          "Por enquanto, pulamos.")


DS_SIT_TOT_TURNO ainda não tem informação real (a eleição de 2026 não ocorreu) — essa seção fica pronta pra usar assim que a Fase 0 for rodada de novo com o resultado oficial. Por enquanto, pulamos.


---
## Conclusão: A Nomeação dos Clusters Faz Sentido?

A mineração de regras de associação pelo algoritmo Apriori responde com **clareza matemática absoluta: SIM, os nomes atribuídos aos clusters na Fase 2 fazem total sentido**.

As 4 variáveis contínuas (Idade, Anos de Estudo, Total de Bens e Total de Despesas Contratadas) foram transformadas em **quartis** para integração na cesta do Apriori:
- **Idade**: 4 quartis puros via `pd.qcut`.
- **Anos de Estudo**: 4 faixas alinhadas aos graus de instrução reais (Até Fundamental, Ensino Médio, Superior Incompleto, Superior Completo).
- **Total de Bens e Total de Despesas**: **quartis *zero-aware*** — os zeros (43% e 26% da base, respectivamente) foram separados como categoria própria e os quartis calculados sobre valores positivos, evitando colapso de faixas.

Ao cruzar os itens frequentes, as regras de associação demonstraram que cada nome reflete de forma fidedigna as condições necessárias e as características determinantes dos candidatos:

### 1. 🏛️ Estruturados (Confiança: 100% | Lift: 1.90)
* **Regras Determinantes:**
  `Bens Q3 (alto) + Despesa Q3 (alta) → Estruturados` (Confiança: 100% | n = 282)
  `Bens Q2 (médio) + Despesa Q3 (alta) → Estruturados` (Confiança: 100% | n = 142)
* **DNA revelado (cluster → características):** Deputados (lift 1,83), Despesa Q3 (lift 1,73), Bens Q3 (lift 1,70), partidos PP e PT.
* **Validação:** Candidatos que combinam patrimônio e gasto representam a máquina eleitoral institucional.

### 2. ⚡ Financiados (Confiança: 100% | Lift: 4.76)
* **Regras Determinantes:**
  `Sem Bens + Despesa Q3 (alta) → Financiados` (Confiança: 100% | n = 49)
  `Sem Bens + Despesa Q2 (média) → Financiados` (Confiança: 100% | n = 167)
* **DNA revelado:** Sem Bens (lift 2,80), Despesa Q1 (lift 2,11), MISSÃO (lift 2,06), **Feminino (51,5%, lift 1,37)**.
* **Validação:** Candidatos sem patrimônio que recebem aportes partidários — perfil evidência das cotas de financiamento feminino.

### 3. 🌱 Base Simbólica (Confiança: 100% | Lift: 6.67)
* **Regra Determinante:**
  `Sem Bens + Sem Despesa → Base Simbólica` (Confiança: 100% | Lift: 6,67 | n = 308 — 94% do cluster!)
* **DNA revelado:** DC (lift 4,51), Sem Despesas (lift 3,82), Sem Bens (lift 2,74), Até Fundamental (lift 1,93), Ensino Médio (lift 1,75).
* **Validação:** Candidaturas formais sem capital financeiro nem mobilização econômica.

### 4. 💼 Patrimônio Sem Campanha (Confiança: 100% | Lift: 8.83 — maior da base!)
* **Regras Determinantes:**
  `Bens Q2 (médio) + Sem Despesa → Patrimônio Sem Campanha` (Confiança: 100% | Lift: 8,83 | n = 63)
  `Bens Q3 (alto) + Sem Despesa → Patrimônio Sem Campanha` (Confiança: 100% | Lift: 8,83 | n = 48)
* **DNA revelado:** Sem Despesas (lift 3,77), Bens Q1 (lift 2,48), Masculino (lift 1,22), Idade Q4 (lift 1,14).
* **Validação:** Candidatos com patrimônio que optaram por não investir na campanha — advogados, empresários, grupo mais idoso e masculino.